app/src/main/java/agency/five/codebase/android/movieapp/data/repository/MovieRepositoryImpl.kt

```diff
@@ -0,0 +1,112 @@
+                }
+                emit(movieResponse.movies)
+            }.flatMapLatest { apiMovies ->
+                movieDao.favorites()
+                    .map { favoriteMovies ->
+                        apiMovies.map { apiMovie ->
+                            apiMovie.toMovie(isFavorite = favoriteMovies.any { it.id == apiMovie.id })
+                        }
+                    }
+            }.shareIn(
+                scope = CoroutineScope(bgDispatcher),
+                started = SharingStarted.WhileSubscribed(1000L),
+                replay = 1,
+            )
+        }
+
+    private val favorites = movieDao.favorites().map {
+        it.map { dbFavoriteMovie ->
+            Movie(
+                id = dbFavoriteMovie.id,
+                imageUrl = dbFavoriteMovie.posterUrl,
+                title = "",
+                overview = "",
+                isFavorite = true,
+            )
+        }
+    }.shareIn(
+        scope = CoroutineScope(bgDispatcher),
+        started = SharingStarted.WhileSubscribed(1000L),
+        replay = 1,
+    )
+
+    override fun movies(movieCategory: MovieCategory): Flow<List<Movie>> =
+        moviesByCategory[movieCategory]!!
+
+    override fun movieDetails(movieId: Int): Flow<MovieDetails> = flow {
+        emit(movieService.fetchMovieDetails(movieId) to movieService.fetchMovieCredits(movieId))
+    }.flatMapLatest { (apiMovieDetails, apiMovieCredits) ->
+        movieDao.favorites()
+            .map { favoriteMovies ->
+                apiMovieDetails.toMovieDetails(
+                    isFavorite = favoriteMovies.any { it.id == apiMovieDetails.id },
+                    crew = apiMovieCredits.crew,
+                    cast = apiMovieCredits.cast
+                )
+            }
+    }.flowOn(bgDispatcher)
+
+    override fun favoriteMovies(): Flow<List<Movie>> = favorites
+
+    override suspend fun addMovieToFavorites(movieId: Int) {
+        val movie = findMovie(movieId)
+        movie?.imageUrl?.let { movieDao.insertMovie(DbFavoriteMovie(movie.id, it)) }
+    }
+
+    override suspend fun removeMovieFromFavorites(movieId: Int) = movieDao.deleteMovie(movieId)
+
+    override suspend fun toggleFavorite(movieId: Int) {
+        runBlocking(bgDispatcher) {
+            val favoriteMovies = favorites.first()
+            if (favoriteMovies.any { it.id == movieId }) {
+                removeMovieFromFavorites(movieId)
+            } else {
+                addMovieToFavorites(movieId)
+            }
+        }
+    }
+
+    private suspend fun findMovie(movieId: Int): Movie {
+        lateinit var movie: Movie
+        moviesByCategory.values.forEach { value ->
+            val movies = value.first()
+            movies.forEach {
+                if (it.id == movieId) {
+                    movie = it
+                    return movie
+                }
+            }
+        }
+        return movie
```

Comment:
No need for `movie` variable as you're returning the movie if you found one. In case there is no movie with provided `movieId`, you can return null or throw exception